File: JavaScript/Data Structure/Tree/binarySearchTree.js
class Node {
    constructor(value) {
        this.left = null;
        this.right = null;
        this.value = value;
    }
}

class BinarySearchTree {
    constructor() {
        this.root = null;
    }
    insert(value) {
        if (!this.root) {
            this.root = new Node(value);
            return;
        } else {
            this._insert(this.root, value);
        }
    }

    _insert(node, value) {
        if (value < node.value) {
            if (!node.left) {
                node.left = new Node(value);
            } else {
                this._insert(node.left, value);
            }
        } else {
            if (!node.right) {
                node.right = new Node(value);
            } else {
                this._insert(node.right, value);
            }
        }
    }

    search(value) {
        if (!this.root) {
            return null;
        }
        let currentNode = this.root;

        while (currentNode) {
            if (value < currentNode.value) {
                currentNode = currentNode.left;
            } else if (value > currentNode.value) {
                currentNode = currentNode.right
            } else if (value === currentNode.value) {
                return currentNode.value;
            } else {
                return null;
            }
        }
    }

    breadthFirstSearch() {
        if (!this.root) {
            return null;
        }
        let currentNode = this.root;
        let list = [];
        let queue = [currentNode];

        while (queue.length > 0) {
            currentNode = queue.pop();
            list.push(currentNode.value);

            if (currentNode.left) {
                queue.push(currentNode.left);
            }
            if (currentNode.right) {
                queue.push(currentNode.right);
            }
        }
        return list;
    }

    dfsInOrder() {
        return this.traverseInOrder(this.root, []);
    }

    dfsPreOrder() {
        return this.traversePreOrder(this.root, []);
    }

    dfsPostOrder() {
        return this.traversePostOrder(this.root, []);
    }

    traverseInOrder(node, list) {
        if (node.left) {
            this.traverseInOrder(node.left, list);
        }

        list.push(node.value);

        if (node.right) {
            this.traverseInOrder(node.right, list);
        }
        return list;
    }

    traversePreOrder(node, list) {
        list.push(node.value);

        if (node.left) {
            this.traverseInOrder(node.left, list);
        }

        if (node.right) {
            this.traverseInOrder(node.right, list);
        }
        return list;
    }

    traversePostOrder(node, list) {
        if (node.left) {
            this.traverseInOrder(node.left, list);
        }

        if (node.right) {
            this.traverseInOrder(node.right, list);
        }

        list.push(node.value);
        return list;
    }

    sums(currentNode = this.root) {
        if (!currentNode) {
            return 0;
        }
        const leftSum = this.sums(currentNode.left);
        const rightSum = this.sums(currentNode.right);

        return currentNode.value + leftSum + rightSum;
    }

    depth(value) {
        return this._depth(this.root, value, 1);
    }

    _depth(node, value, depth) {
        if (!node) {
            return -1;
        }
        if (node.value === value) {
            return depth;
        }
        if (value < node.value) {
            return this._depth(node.left, value, depth + 1);
        } else {
            return this._depth(node.right, value, depth + 1);
        }
    }

    height() {
        return this._height(this.root);
    }

    _height(node) {
        if (!node) {
            return -1;
        }
        const leftBranch = this._height(node.left);
        const rightBranch = this._height(node.right);

        return 1 + Math.max(leftBranch, rightBranch);
    }

}

const bst = new BinarySearchTree();

bst.insert('C');
bst.insert('JavaScript');
bst.insert('Python');
bst.insert('Odin');
bst.insert('Haskell');

console.log(bst);

console.log(bst.breadthFirstSearch());
console.log(bst.dfsInOrder());

console.log(bst.depth('Python'));

console.log(bst.height());

File: Go/Data Structure/Tree/binarySearchTree.go
package main

import "fmt"

type Node struct {
	left  *Node
	right *Node
	value string
}

type BinarySearchTree struct {
	root *Node
}

func NewBST() *BinarySearchTree {
	return &BinarySearchTree{}
}

func (bst *BinarySearchTree) Insert(value string) {
	if bst.root == nil {
		bst.root = &Node{value: value}
	} else {
		bst._insert(bst.root, value)
	}
}

func (bst *BinarySearchTree) _insert(node *Node, value string) {
	if len(value) < len(node.value) {
		if node.left == nil {
			node.left = &Node{value: value}
		} else {
			bst._insert(node.left, value)
		}
	} else {
		if node.right == nil {
			node.right = &Node{value: value}
		} else {
			bst._insert(node.right, value)
		}
	}
}

func (bst *BinarySearchTree) Search(value string) string {
	currentNode := bst.root
	for currentNode != nil {
		if value < currentNode.value {
			currentNode = currentNode.left
		} else if value > currentNode.value {
			currentNode = currentNode.right
		} else {
			return currentNode.value
		}
	}
	return ""
}

func (bst *BinarySearchTree) BreadthFirstSearch() []string {
	if bst.root == nil {
		return nil
	}
	var list []string = []string{}
	var queue []*Node = []*Node{bst.root}

	for len(queue) > 0 {
		currentNode := queue[0]
		queue = queue[1:]
		list = append(list, currentNode.value)

		if currentNode.left != nil {
			queue = append(queue, currentNode.left)
		}
		if currentNode.right != nil {
			queue = append(queue, currentNode.right)
		}
	}
	return list
}

func (bst *BinarySearchTree) DFSInOrder() []string {
	return bst.traverseInOrder(bst.root, []string{})
}

func (bst *BinarySearchTree) traverseInOrder(node *Node, list []string) []string {
	if node == nil {
		return list
	}
	list = bst.traverseInOrder(node.left, list)
	list = append(list, node.value)
	list = bst.traverseInOrder(node.right, list)
	return list
}

func (bst *BinarySearchTree) DFSPreOrder() []string {
	return bst.traversePreOrder(bst.root, []string{})
}

func (bst *BinarySearchTree) traversePreOrder(node *Node, list []string) []string {
	if node == nil {
		return list
	}
	list = append(list, node.value)
	list = bst.traversePreOrder(node.left, list)
	list = bst.traversePreOrder(node.right, list)
	return list
}

func (bst *BinarySearchTree) DFSPostOrder() []string {
	return bst.traversePostOrder(bst.root, []string{})
}

func (bst *BinarySearchTree) traversePostOrder(node *Node, list []string) []string {
	if node == nil {
		return list
	}
	list = bst.traversePostOrder(node.left, list)
	list = bst.traversePostOrder(node.right, list)
	list = append(list, node.value)
	return list
}

func (bst *BinarySearchTree) DepthFirstSearch() []string {
	stack := []*Node{}
	current := bst.root
	var result []string

	for current != nil || len(stack) > 0 {
		for current != nil {
			stack = append(stack, current)
			current = current.left
		}

		current = stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		result = append(result, current.value)

		current = current.right
	}
	return result
}

func (bst *BinarySearchTree) Sums() int {
	return bst.sums(bst.root)
}

func (bst *BinarySearchTree) sums(node *Node) int {
	if node == nil {
		return 0
	}
	leftSum := bst.sums(node.left)
	rightSum := bst.sums(node.right)
	return len(node.value) + leftSum + rightSum
}

func (bst *BinarySearchTree) FindMax() (string, string) {
	var parent *Node
	node := bst.root.right

	for node.right != nil {
		parent = node
		node = node.right
	}

	if parent == nil {
		return node.value, ""
	}

	return node.value, parent.value
}

func main() {
	bst := NewBST()

	bst.Insert("C")
	bst.Insert("JavaScript")
	bst.Insert("Python")
	bst.Insert("Odin")
	bst.Insert("Haskell")
	bst.Insert("Machine Code")
	bst.Insert("Nim")
	bst.Insert("Ruby")

	fmt.Println(bst.Search("Haskell"))
	fmt.Println(bst.BreadthFirstSearch())
	fmt.Println(bst.DFSInOrder())
	fmt.Println(bst.DFSPreOrder())
	fmt.Println(bst.DFSPostOrder())
	fmt.Println(bst.Sums())

	fmt.Println(bst.DepthFirstSearch())

	maxValue, parentValue := bst.FindMax()
	fmt.Println(maxValue, parentValue)
}
